### PR TITLE
Tracking evaded conjunctions

### LINKED ISSUE #133: Track # of conjunctions which were evaded
# Feature
## Desired behavior
We should track these as a sanity check and to know about potential CDMs. 

## What needs to be done
Store conjunctions even if evaded

## How it can be tested
Run some test case?

### PR DESCRIPTION
**Pull request by gomezzz**
# Description

- Added tracking of evaded collisions
- Added IO for evaded collisions
- Updated HDF5 tests
- Small fix in tests
- Temporarily disabled one check in regularGrid decomp test as broken

## Related Pull Requests

- #150 

## Resolved Issues

- [x] fixes #133 

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] Running CI

### CODE REVIEW

File: tests/mpitestladds/distributedMemParallelization/RegularGridDecompositionTests.cpp [hunk 1]
```diff
@@ -240,17 +241,19 @@ TEST_F(RegularGridDecompositionTests, testGridDecompCollisions) {
               << autopas::utils::ArrayUtils::to_string(p.getPosition()) << " is arriving." << std::endl;
 
   // Check if the arriving particles are correct
-  if (rank == 7)
-    ASSERT_EQ(incomingParticles.size(), 2) << "Expected " << 2 << " particles on rank " << rank;
-  else
-    ASSERT_EQ(incomingParticles.size(), 0) << "Expected " << 0 << " particles on rank " << rank;
+  // TODO currently there is a bug with comms for this decomposition, reenable when fixed
```
Comment (FG-TUM) [inline tests/mpitestladds/distributedMemParallelization/RegularGridDecompositionTests.cpp:244]: don't disable via comment. Use `GTEST_SKIP_(message)`

**Review by FG-TUM — CHANGES_REQUESTED** [review]
some small stuff

File: src/ladds/CollisionFunctor.cpp [hunk 2]
```diff
@@ -152,6 +147,7 @@ void CollisionFunctor::SoAFunctorPair(autopas::SoAView<SoAArraysType> soa1,
 #pragma omp declare reduction(vecMerge:CollisionCollectionT \
                               : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()))
     // alias because OpenMP needs it
+    // TODO add evadedCollisions when refactoring this
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply src/ladds/CollisionFunctor.cpp:150]: SoA functor is currently not operational, so I don't see any point in (nor way of)  fixing it right now. However, when / if we do we will need to add this here

File: src/ladds/CollisionFunctor.cpp [hunk 3]
```diff
@@ -219,6 +215,7 @@ void CollisionFunctor::SoAKernel(
   }
 
   // store pointers to colliding pair
+  // TODO add evaded collisions when refactoring this
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply src/ladds/CollisionFunctor.cpp:218]: (as above)

File: src/ladds/utils/DistanceApproximation.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  // If in the past, minimum is at t = 0
+  // Else If in future timesteps, minimum for this is at t = dt
+  t = std::clamp(t, 0., dt);
+
+  // Compute actual distance by propagating along the line to t
+  const auto p1 = add(old_r_i, mulScalar(vi, t));
+  const auto p2 = add(old_r_j, mulScalar(vj, t));
+
+  return std::make_tuple(sub(p1, p2), p2);
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply src/ladds/utils/DistanceApproximation.h:64]: actually those are the particle positions at t not distances. will change according to that instead

File: tests/mpitestladds/distributedMemParallelization/RegularGridDecompositionTests.cpp (hunk 1 above)
Comment (gomezzz) [reply tests/mpitestladds/distributedMemParallelization/RegularGridDecompositionTests.cpp:244]: The rest of the test is functional though and useful things are tested? I'd only want to skip this assertion?

File: src/ladds/CollisionFunctor.cpp [hunk 5]
```diff
@@ -12,20 +12,28 @@
 #include <algorithm>
 #include <limits>
 
+#include "ladds/utils/DistanceApproximation.h"
+
 namespace LADDS {
-CollisionFunctor::CollisionFunctor(double cutoff, double dt, double collisionDistanceFactor, double minDetectionRadius)
+CollisionFunctor::CollisionFunctor(
+    double cutoff, double dt, double collisionDistanceFactor, double minDetectionRadius, double CDMcutoffInKM)
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply src/ladds/CollisionFunctor.cpp:19]: renamed to evasionTrackingCutoffInKM

File: cfg/default_cfg.yaml [hunk 6]
```diff
@@ -11,6 +11,7 @@ sim:
   minAltitude: 150 # Everything below this altitude above ground will be considered burning up [km]
   deltaT: 10.0 # [s]
   collisionDistanceFactor: 1.0 # Factor multiplied with the sum of radii to over approximate collision distances.
+  # CDMcutoffInKM: 0.1 # Distance at which even evaded conjunctions are tracked (in a separate out file)
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply cfg/default_cfg.yaml]: renamed

File: tests/mpitestladds/distributedMemParallelization/RegularGridDecompositionTests.cpp (hunk 1 above)
Comment (gomezzz) [reply tests/mpitestladds/distributedMemParallelization/RegularGridDecompositionTests.cpp:244]: done

**Review by FG-TUM — APPROVED** [review]
I do not approve of the commit messages :stuck_out_tongue_winking_eye:

**Pull request merged**